compiler/rustc_builtin_macros/src/deriving/generic/mod.rs

```diff
@@ -1532,11 +1579,76 @@ impl<'a> TraitDef<'a> {
                 let mut exprs: Vec<_> = mk_exprs(i, struct_field, sp);
                 let self_expr = exprs.remove(0);
                 let other_selflike_exprs = exprs;
+                let mut skipped_derives = SkippedDerives::None;
+                let skip_enabled = cx.ecfg.features.derive_skip
+                    || struct_field.span.allows_unstable(sym::derive_skip);
+                for attr in attr::filter_by_name(&struct_field.attrs, sym::skip) {
+                    if !skip_enabled {
+                        rustc_session::parse::feature_err(
+                            &cx.sess,
+                            sym::derive_skip,
+                            attr.span,
+                            "the `#[skip]` attribute is experimental",
+                        )
+                        .emit();
+                    }
+                    let Some(skip_attr) = ast::Attribute::meta_kind(attr) else {
+                        unreachable!()
+                    };
+
+                    // FIXME: better errors
+                    match skip_attr {
+                        ast::MetaItemKind::Word => {
+                            skipped_derives = SkippedDerives::All;
+                            break;
+                        }
+                        ast::MetaItemKind::List(items) => {
+                            for item in items {
+                                let span = item.span();
+                                let ast::NestedMetaItem::MetaItem(ast::MetaItem {
+                                    path,
+                                    kind: ast::MetaItemKind::Word,
+                                    ..
+                                }) = item
+                                else {
+                                    cx.dcx().emit_err(errors::DeriveSkipBadArgument {
+                                        span,
+                                    });
+                                    continue;
+                                };
+                                let name = path.segments[0].ident;
+                                const SUPPORTED_TRAITS: [Symbol; 5] = [
+                                    sym::PartialEq,
+                                    sym::PartialOrd,
+                                    sym::Ord,
+                                    sym::Hash,
+                                    sym::Debug,
+                                ];
```

Comment:
Not sure how I feel about hard-coding the list of supported traits. Technically speaking a built-in derive macro supports `#[skip]` if it declares `skip` as a helper attribute and conversely if it doesn't declare it then it doesn't support it. I wonder if we could somehow leverage this information (if it's available)?